### DISCUSSION TITLE
The "EventEmitter memory leak" issue with Node Test Runner

### DISCUSSION THREAD

**Discussion by sernaferna** [nodejs/help#4669] [Q&A]
_This might be more prevalent on Windows than other operating systems?_

I get that **watch mode** is experimental so I can't expect it to work perfectly but I'm wondering if anyone has tips or tricks for getting it to work _more_ consistently? 🙂 

I find myself regularly kicking off my tests using combinations of `--watch` and `--test-only` and immediately having to _Ctrl+c_ and restart until it starts a "clean" execution. Sometimes it works well right off the bat; sometimes I multitask and work on something else, constantly Alt+Tabbing back to VS Code to _Ctrl+C_ and restart the tests over and over again until it works.

I'll note that I'm still on Node 20 because I want to be on the LTS version.

**Comment by avivkeller** [comment]
@nodejs/test_runner

**Reply by sernaferna** [reply]
I'm sensing this means I've posted to the wrong discussion group, but I can't find the test_runner one you're referring to; do you have a link?

**Reply by avivkeller** [reply]
No, it doesn’t mean that. I’ve pinged the team of people who have a better understanding of the issue at hand.

**Reply by sernaferna** [reply]
Ah, even better!

**Comment by juliangruber** [comment]
> I'll note that I'm still on Node 20 because I want to be on the LTS version.

Have you tried reproducing this issue on Node 22?

**Comment by sernaferna** [comment]
Quick note to say that I haven't found the time to try Node 22 yet, _however,_ I have found something that helps (that maybe was obvious but I missed): setting `NODE_ENV` to `test` makes things much more stable. (In my case I have a function I use to execute my tests programmatically, so I set it right at the beginning of that before calling `run()` from `node:test`.)

This seems to work _much_ more reliably for most of the applications in my monorepo. (I have a React app for which I still get a ton of `MaxListenersExceededWarning` / `EventEmitter memory leak detected` issues, and I'm not sure why I'm having more issues with it than with the others, but for now I'm just avoiding **watch mode** when testing that app unless I have to...)

**Reply by sernaferna** [reply]
Apologies for the spam, but correction: Despite an afternoon in which everything seemed to be very stable, I haven't subsequently seen any change to stability, even with `NODE_ENV` set to `test`. So I'll have to set aside an afternoon someday when I have time to change everything to Node 22 and see what happens (before changing everything back).

**Comment by benjamingr** [comment]
I use the test runner and didn't run into this issue. If you make a self contained repro (a gh repo with minimal code to reproduce the issue) I am happy to take a look and try and fix whatever is causing this. 

Bug reports are valuable :]
